ext/regex.go

```diff
@@ -330,3 +334,107 @@ func extractAll(target, regexStr ref.Val) ref.Val {
+			// The regex cost is the size of the regex pattern, scaled by a complexity factor.
+			regexCost := estimateSize(c, args[1]).Add(checker.FixedSizeEstimate(1)).MultiplyByCostFactor(common.RegexStringLengthCostFactor)
+
+			// The result is a list of strings. Worst Case: its contents are the size of the entire target.
+			resultSize := &checker.SizeEstimate{Min: 0, Max: targetSize.Max}
+			// The cost to allocate the result list is its base cost plus the size of its contents.
+			listAllocCost := resultSize.Add(checker.FixedSizeEstimate(common.ListCreateBaseCost)).MultiplyByCostFactor(1)
+			// The total cost is the search cost (target * regex) plus the allocation cost for the result list.
+			return &checker.CallEstimate{
+				CostEstimate: targetCost.Multiply(regexCost).Add(listAllocCost),
+				ResultSize:   resultSize,
+			}
+		}
+		return nil
+	}
+}
+
+func estimateReplaceCost() checker.FunctionEstimator {
+	return func(c checker.CostEstimator, target *checker.AstNode, args []checker.AstNode) *checker.CallEstimate {
+		l := len(args)
+		if l == 3 || l == 4 {
+			targetSize := estimateSize(c, args[0])
+			replacementSize := estimateSize(c, args[2])
+			// Fixed size estimate of +1 is added for safety from zero size args.
+			// The target cost is the size of the target string, scaled by a traversal factor.
+			targetCost := targetSize.Add(checker.FixedSizeEstimate(1)).MultiplyByCostFactor(common.StringTraversalCostFactor)
+			// The regex cost is the size of the regex pattern, scaled by a complexity factor.
+			regexCost := estimateSize(c, args[1]).Add(checker.FixedSizeEstimate(1)).MultiplyByCostFactor(common.RegexStringLengthCostFactor)
+			// Estimate the build cost of the new string using a pessimistic upper bound.
+			replacementSizeEstimate := targetSize.Multiply(replacementSize.Add(checker.FixedSizeEstimate(1)))
+
+			// Estimate the potential size range of the output string. The final size could be smaller
+			// (if the replacement size is 0) or larger than the original.
+			allReplacedSize := targetSize.Max * replacementSize.Max
+			noneReplacedSize := targetSize.Max
+			resultSize := &checker.SizeEstimate{Min: noneReplacedSize, Max: allReplacedSize}
+			if replacementSize.Max == 0 {
+				resultSize = &checker.SizeEstimate{Min: allReplacedSize, Max: noneReplacedSize}
+			}
+			// The total cost is the search cost (target * regex) plus the estimated build cost.
+			return &checker.CallEstimate{
+				CostEstimate: targetCost.Multiply(regexCost).Add(checker.CostEstimate(replacementSizeEstimate)),
+				ResultSize:   resultSize,
+			}
+		}
+		return nil
+	}
+}
+
+func trackReplaceCost() interpreter.FunctionTracker {
+	return func(args []ref.Val, result ref.Val) *uint64 {
+		targetSize := float64(actualSize(args[0])) * float64(common.StringTraversalCostFactor)
+		regexSize := float64(actualSize(args[1])) * float64(common.RegexStringLengthCostFactor)
```

Comment:
the units are a bit confusing here, consider renaming targetCost and regexCost

size * (cost / size) -> cost

I think the tracking estimate should be close to the estimator for 'matches' in [runtimecost.go](https://source.corp.google.com/piper///depot/google3/third_party/cel/go/interpreter/runtimecost.go)

plus the cost of creating the result string

```
let targetCost = ceil( float(1 + target_size) * StringTraversalFactor)
let regexCost = ceil(float(pattern_size) * RegexLengthFactor)
let resultCost = ceil(float(result_size) * StringTraversalFactor)

let estimate = targetCost*regexCost + resultCost + callCost
```